ruby/ql/test/library-tests/dataflow/array-flow/array_flow.rb

```diff
@@ -0,0 +1,523 @@
+end
+
+def m36(i)
+    a = [0, 1, source(36.1), source(36.2)]
+    b = a.drop(i)
+    sink(b[0]) # $ hasValueFlow=36.1 # $ hasValueFlow=36.2
+    b = a.drop(1)
+    sink(b[0])
+    sink(b[1]) # $ hasValueFlow=36.1
+    sink(b[i]) # $ hasValueFlow=36.1 # $ hasValueFlow=36.2
+    a[i] = source(36.3)
+    b = a.drop(1)
+    sink(b[1]) # $ hasValueFlow=36.1 # $ hasValueFlow=36.3
+    c = b.drop(100)
+    sink(c[1]) # $ hasValueFlow=36.3
+end
+
+def m37
+    a = [0, 1, source(37.1), source(37.2)]
+    b = a.drop_while do |x|
+        sink x # $ hasValueFlow=37.1 # $ hasValueFlow=37.2
+    end
+    sink(b[0]) # $ hasValueFlow=37.1 # $ hasValueFlow=37.2
+end
+
+def m38
+    a = [0, 1, source(38)]
+    b = a.each do |x|
+        sink x # $ hasValueFlow=38
+    end
+    sink(b[2]) # $ hasValueFlow=38
+end
+
+def m39
+    a = [0, 1, source(39)]
+    b = for x in a # desugars to an `each` call
+        sink x # $ hasValueFlow=39
+    end
+    sink x # $ hasValueFlow=39
+    sink(b[2]) # $ hasValueFlow=39
+end
+
+def m40
+    a = [0, 1, source(40)]
+    a.each_cons(2) do |x|
+        sink (x[0]) # $ hasValueFlow=40
+    end
+end
+
+def m41
+    a = [0, 1, source(41)]
+    b = a.each_entry do |x|
+        sink x # $ hasValueFlow=41
+    end
+    sink(b[2]) # $ hasValueFlow=41
+end
+
+def m42
+    a = [0, 1, source(42)]
+    b = a.each_index do |x|
+        sink x
+    end
+    sink(b[2]) # $ hasValueFlow=42
+end
+
+def m43
+    a = [0, 1, 2, source(43)]
+    b = a.each_slice do |x|
+        sink(x[0]) # $ hasValueFlow=43
+    end
+    sink(b[3]) # $ hasValueFlow=43
+end
+
+def m44
+    a = [0, 1, 2, source(44)]
+    b = a.each_with_index do |x,i|
+        sink(x) # $ hasValueFlow=44
+        sink(i)
+    end
+    sink(b[3]) # $ hasValueFlow=44
+end
+
+def m45
+    a = [0, 1, 2, source(45.1)]
+    b = a.each_with_object(source(45.2)) do |x,a|
+        sink(x) # $ hasValueFlow=45.1
+        sink(a) # $ hasValueFlow=45.2
+    end
+    sink(b) # $ hasValueFlow=45.2
+end
+
+def m46(i)
+    a = [0, 1, 2, source(46.1)]
+    b = a.fetch(source(46.2)) do |x|
+        sink(x) # $ hasValueFlow=46.2
+    end
+    sink(b) # $ hasValueFlow=46.1
+end
+
+def m47
+    a = [0, 1, 2, source(47.1)]
+    a.fill(source(47.2), 1, 1)
+    sink(a[3]) # $ hasValueFlow=47.1 $ hasValueFlow=47.2
+    a.fill(source(47.3))
+    sink(a[0]) # $ hasValueFlow=47.3
+    a.fill do |i|
+        source(47.4)
+    end
+    sink(a[0]) # $ hasValueFlow=47.4
+    a.fill(2) do |i|
+        source(47.5)
+    end
+    sink(a[0]) # $ hasValueFlow=47.4 $ hasValueFlow=47.5
+end
+
+def m48
+    a = [0, 1, 2, source(48)]
+    b = a.filter do |x|
+        sink(x) # $ hasValueFlow=48
+    end
+    sink(b[0]) # $ hasValueFlow=48
+end
+
+def m49
+    a = [0, 1, 2, source(49)]
+    b = a.filter_map do |x|
+        sink(x) # $ hasValueFlow=49
+    end
+    sink(b[0]) # $ hasValueFlow=49
+end
+
+def m50
+    a = [0, 1, 2, source(50)]
+    b = a.filter! do |x|
+        sink(x) # $ hasValueFlow=50
```

Comment:
`Array#filter!` only returns a result if some elements are removed, and it returns the value of the array after the filter is applied. If no elements are removed, it returns `nil`. So e.g.
```rb
[1,2,3].filter { true } # => nil
[1,2,3].filter { false } # => []
[1,2,3].filter { |x| x > 2 } # => [3]
```

Not really important, but for the source to reach the sink in reality this block would need something like `x > 2` at the end.